File: src/services/validate.ts
import { config } from "../config.js";

const SUSPICIOUS_RE = /[<>"'`;$|&{}\\]/;
const YT_ID_RE = /^[a-zA-Z0-9_-]{11}$/;

export interface ValidationError {
  valid: false;
  reason: string;
}

export interface ValidationSuccess {
  valid: true;
}

type ValidationResult = ValidationError | ValidationSuccess;

export function validateQuery(query: string): ValidationResult {
  if (!query || query.trim().length === 0) {
    return { valid: false, reason: "Query is required" };
  }
  if (query.length > config.maxQueryLength) {
    return { valid: false, reason: `Query exceeds ${config.maxQueryLength} characters` };
  }
  if (SUSPICIOUS_RE.test(query)) {
    return { valid: false, reason: "Query contains invalid characters" };
  }
  return { valid: true };
}

export function validateTrackId(id: string): ValidationResult {
  if (!id || id.trim().length === 0) {
    return { valid: false, reason: "Track ID is required" };
  }
  if (id.length > 100) {
    return { valid: false, reason: "Track ID too long" };
  }
  if (SUSPICIOUS_RE.test(id)) {
    return { valid: false, reason: "Track ID contains invalid characters" };
  }
  return { valid: true };
}

File: src/services/validate.py
from __future__ import annotations

import re

from src.config import settings

SUSPICIOUS_RE = re.compile(r'[<>"\'`;$|&{}\\]+')


def validate_query(query: str) -> str | None:
    if not query or not query.strip():
        return "Query is required"
    if len(query) > settings.max_query_length:
        return f"Query exceeds {settings.max_query_length} characters"
    if SUSPICIOUS_RE.search(query):
        return "Query contains invalid characters"
    return None


def validate_track_id(track_id: str) -> str | None:
    if not track_id or not track_id.strip():
        return "Track ID is required"
    if len(track_id) > 100:
        return "Track ID too long"
    if SUSPICIOUS_RE.search(track_id):
        return "Track ID contains invalid characters"
    return None
